### PR TITLE
fix limitations on applying acls when scheme/authority is given in path

### PR DESCRIPTION
**Pull request by zzeekk**
Bug was in getPathLevel. Custom implementation is now replaced with hadoop default implementation which should be robust.

Additional changes:
Custom Implementation of parent replaced with hadoop default implementation.
Added additional test cases for paths with schema/authority.
Converted path:String parameters to path:Path to have a better definition.
Remove unneeded Options in method declarations.
Implemented user home check more generic.
Extracted configuration parameters to Environment object.
Forced validation of hdfsAclsMinLevelPermissionOverwrite on start of addACLs.
Removed unused methods.
Changed test cases to methods which are actually in use (traverseDirectoryUp).
Improved documentation.

### CODE REVIEW

File: src/main/scala/io/smartdatalake/definitions/Environment.scala [hunk 1]
```diff
@@ -30,6 +30,24 @@ object Environment {
...
   var hdfsAclsRequired: Boolean = false
 
+  /**
+   * Modifying ACL's is only allowed up to the following level (default=2)
+   * See also [[io.smartdatalake.util.misc.AclUtil]]
+   */
+  var hdfsAclsMinLevelPermissionModify = 2 // up to user home
+
+  /**
+   * Overwriting ACL's is only allowed up to the following level (default=2)
```
Comment (pgruetter) [inline src/main/scala/io/smartdatalake/definitions/Environment.scala:40]:
```suggestion
   * Overwriting ACL's is only allowed up to the following level (default=5)
```
I assume our Environment reflects the "default".

File: src/main/scala/io/smartdatalake/definitions/Environment.scala [hunk 2]
```diff
@@ -30,6 +30,24 @@ object Environment {
...
+   * See also [[io.smartdatalake.util.misc.AclUtil]]
+   */
+  var hdfsAclsMinLevelPermissionOverwrite = 5 // incl. and underneath feed
+
+  /**
+   * Limit setting ACL's to user home (default=true)
+   */
```
Comment (pgruetter) [inline src/main/scala/io/smartdatalake/definitions/Environment.scala:47]:
```suggestion
   Declare on which level in the directory hierarchy your user homes are (usually on level 2)
   */
```

File: src/main/scala/io/smartdatalake/util/misc/AclUtil.scala [hunk 3]
```diff
@@ -48,60 +49,57 @@ private[smartdatalake] case class AclElement (aclType:String, name:String, permi
...
+ * Levels are counted as follows
+ * Root hdfs://nameservice1/ or also "/" -> 0
+ * User Dir on Hadoop hdfs://nameservice1/user -> 1
+ * User Home hdfs://nameservice1/user/app_dir -> 2
+ * Layer hdfs://nameservice1/user/app_dir/<layer> -> 3
+ * Source hdfs://nameservice1/user/app_dir/<layer>/<source_name> -> 4
+ * Feed hdfs://nameservice1/user/app_dir/<layer>/<source_name>/<feed_name> -> 5
+ * ...
+ *
+ * In this structure good limitations are: MinLevelPermissionModify=2, MinLevelPermissionOverwrite=5
```
Comment (pgruetter) [inline src/main/scala/io/smartdatalake/util/misc/AclUtil.scala:77]:
```suggestion
 * Utils for setting ACLs on Hadoop Filesystems
 *
 * Setting ACLs in the directory hierarchy is a sensitive task.
 * In SmartDataLake ACLs can be configured for HadoopFileDataObjects.
 * Setting ACLs on a given path of a DataObject has to
 * 1) overwrite ACLs on the directory and its subdirectories
 * 2) modify ACLs on the parent directories up to a certain level configured by hdfsAclMinLevelPermissionModify
 * Additional limitations for overwriting ACLs can be configured by hdfsAclMinLevelPermissionOverwrite
 *
 * By default, changing ACLs is also limited to the user home directory. This is configured by hdfsAclsLimitToUserHome and hdfsAclsUserHomeLevel.
 * It checks if path at hdfsAclsUserHomeLevel ends with username.
 *
 * A common structure for directories on HDFS is as follows:
 * hdfs://nameservice1/user/app_dir/<layer>/<source_name>/<feed_name>
 *
 * Levels are counted as follows
 * Root hdfs://nameservice1/ or also "/" -> 0
 * User Dir on Hadoop hdfs://nameservice1/user -> 1
 * User Home hdfs://nameservice1/user/app_dir -> 2
 * Layer hdfs://nameservice1/user/app_dir/<layer> -> 3
 * Source hdfs://nameservice1/user/app_dir/<layer>/<source_name> -> 4
 * Feed hdfs://nameservice1/user/app_dir/<layer>/<source_name>/<feed_name> -> 5
 * ...
 *
 * In this structure good limitations are: MinLevelPermissionModify=2, MinLevelPermissionOverwrite=5
```

Comment (ursin) [reply src/main/scala/io/smartdatalake/util/misc/AclUtil.scala:77]: `dfs.nameservices` in lieu of `nameservice1`, which could be to SBB specific?

**Comment by zzeekk** [comment]
Environment.hadoopAuthoritiesWithAclsRequired can now be set by Environment variable "SDL_HADOOP_AUTHORITIES_WITH_ACLS_REQUIRED" or java system property "sdl.hadoopAuthoritiesWithAclsRequired" by a comma seperated list of authorities.

**Comment by swissrussdev** [comment]
Good change - doesn't it make sense to implement this as a list of nameservices which are exempt from ACLs? That way the default is for ACL enforcement to be switched on.

**Pull request merged**